### PR TITLE
B. refactor(state): prepare non-finalized chains for shared read-only access

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

As part of #3745, we need to be able to clone read-only access to the best non-finalized chain.

### Designs

Each `Chain` is stored in an `Arc`. The inner chain gets cloned before updates, if needed.

To compensate for the extra `Arc` overhead, I did some refactors to only clone the chains when absolutely necessary.

## Solution

- wrap chain set chains in an Arc

Performance refactors:
- rather than taking the chain we want to modify, leave the Arc in the chain set, then do the clone as late as possible
- clone `Arc`s rather than cloning the inner `Chain`s
- don't pop chain roots of chains we're going to drop anyway

## Review

This is a bit urgent, because it blocks most of the lightwalletd integration tests.

This PR depends on PR #3809.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

- #3745

### CODE REVIEW

**Comment by teor2345** [comment]
@mergifyio update

File: zebra-state/src/service/non_finalized_state/chain.rs [hunk 1]
```diff
@@ -189,18 +189,31 @@ impl Chain {
...
+        if let Err(error) = self.update_chain_tip_with(&block) {
+            // The chain could be in an invalid half-updated state, so clear its data.
+            *self = Chain::new(
+                self.network,
+                sprout::tree::NoteCommitmentTree::default(),
+                sapling::tree::NoteCommitmentTree::default(),
+                orchard::tree::NoteCommitmentTree::default(),
+                HistoryTree::default(),
+                ValueBalance::zero(),
+            );
```
Comment (jvff) [inline zebra-state/src/service/non_finalized_state/chain.rs:208]: This feels weird. I think the previous way having the method take ownership of `self` avoids the risk of misusing a chain after an error. Maybe we can restore that behaviour with a few changes. I played around a bit to see if it was possible, and got this: #3826.

**Review by jvff — APPROVED** [review]

**Review by jvff — APPROVED** [review]

File: zebra-state/src/service/non_finalized_state/chain.rs (hunk 1 above)
Comment (teor2345) [reply zebra-state/src/service/non_finalized_state/chain.rs:208]: You're right, thanks for the simplification!

**Comment by teor2345** [comment]
@mergifyio refresh

**Pull request merged**